apis_core/apis_entities/tables.py

```diff
@@ -83,6 +84,18 @@ class Meta:
                     )
 
         def __init__(self, *args, **kwargs):
+
+            self.base_columns["detail"] = tables.TemplateColumn(
+                template_name='apis_entities/tables_detail_button_template.html',
+                extra_context={'entity_class':entity}
+            )
+
+            if is_authenticated:
+                self.base_columns["edit"] = tables.TemplateColumn(
+                    template_name='apis_entities/tables_edit_button_template.html',
+                    extra_context={'entity_class':entity}
+                )
+
```

Comment:
Is there an advantage to declaring the columns in the `__init__` method rather than in the class itself?

I'll add my main point of reference is [what I did](https://github.com/acdh-oeaw/apis-core-rdf/blob/dev/apis_core/apis_relations/tables.py#L59) in `PropertyTable` after looking at other table of ours + studying the docs, which didn't include many examples of `__init__` in use.

Reply:
I just applied the same conventions that I found in other uses of this implementation in the existing code base (and how it was explained in the docs). 

Again, you made me think about it. I would leave it as it, but actually I am not 100% sure how this part of the code runs. Here are my assumptions, based on my current knowledge:

- the class is nested in a factory function that takes over the creation of the class itself. 
- this matters, I think, because the class itself is constructed by  a special metaclass. What you define within the inner `class Meta` is stuff that will be run when the actual class is constructed (not the instances of the class, which happens in the init).
- usually, metaclasses construct their classes when a python program starts up, and only ONCE. So actually, at this point, no user or request would be present and it would not make sense (I think!) to include the check against the permissions within the `class meta`block. 
- This is bypassed by the fact that the whole class is wrapped in a factory method. This delays the actual construction of the class and also the init of the classes instance(s) to the moment the factory method is called, and so this construction is run each time an instance of the class is requested (which implies that for each request of an instance, the whole class is reconstructed).
- at least, that's my guess. And that would explain why the permission check in the  `class  meta` body actually works in he current implementation.

So I will test my assumptions now. because, if they are correct, this is whole pattern could be improved. As this is sort of defeating the purpose of generic classes and OOP in general. We are creating a new class and a new instance of that class now each time we want to render a single table. As this is not the only place in the codebase where this factory function + metaclass construction pattern occurs, improving on this pattern might be worth it. And it would help with understanding how the code is actually run.

Based on those assumption, to answer your question: 
- I guess in the current pattern it would not matter if the template columns are added in the init or the class body. because, as I said, construction of the class and initialising an instance happen sequentially in this code each time.
- but adding the templates to the init seems still to be more correct to me, as this is stuff should actually be bound to the instance of the table, not the whole class. 
- also, if this pattern might be refactored in the future, having the templates in the class body might break the code or create idiosyncratic behaviour.

Again, this is written from a couple of based assumption that I have about the inner workings of this particular code. Will check them now and update you.

Reply:
I just tested it, my assumptions seem to be correct. 

I copied the code of the table class and pasted it outside the factory function. I then bound this table to a certain entity-model class and added a condition in the view serving the listviews if this entity is requested, to use my TestTable. 
I added print statements in the `__init__` and `class meta` bodies to both implementations and requested the list view of this specific entity and a generic other one in the frontend. 

Here is what django prints: 

```bash
Performing system checks...

meta reached in TestTable class # --> you can see that the TestTable outside the factory is constructed before anything else (this is the moment the class is constructed)

June 01, 2023 - 10:11:52
Django version 4.1.9, using settings 'apis_ontology.settings.server_settings'
Starting development server at http://127.0.0.1:8000/
Quit the server with CONTROL-C.

# Here I request the listview of an entity using the factory function:
meta reached in table class # Class is constructed
init called in table # Instance is initialised
[01/Jun/2023 10:11:58] "GET /apis/entities/entity/e1_crm_entity/list/ HTTP/1.1" 200 13248
[01/Jun/2023 10:11:58] "GET /static/autocomplete_light/select2.js HTTP/1.1" 200 4935
Not Found: /apis/entities/entity/e1_crm_entity/list/piwik.js
[01/Jun/2023 10:11:58] "GET /apis/entities/entity/e1_crm_entity/list/piwik.js HTTP/1.1" 404 1599

# Here I request the entity that uses the table class outside the factory function:
init called in TestTable # --> only an instance is created at runtime, the class is re-used from before startup! 

[01/Jun/2023 10:11:59] "GET /apis/entities/entity/e55_type/list/ HTTP/1.1" 200 7939
[01/Jun/2023 10:11:59] "GET /static/autocomplete_light/select2.js HTTP/1.1" 304 0
Not Found: /apis/entities/entity/e55_type/list/piwik.js
[01/Jun/2023 10:11:59] "GET /apis/entities/entity/e55_type/list/piwik.js HTTP/1.1" 404 1597
```

I think it does no make sense to create a class and an instance of a class each time you want to use a generic class. This is a separate issue, but I already have ideas how to resolve that!

Reply:
@b1rger  @sennierer @koeaw  - This specific thread is important, please have a look. 

Thanks for making me think about it again, @koeaw !

Reply:
So, kay, to answer your question once more: that's why I think placing the template columns in the __init__ block is the correct approach, even if we use the factory function. because the conditions should be applied at runtime each time an instance of the table (listview) is requested, on a per user basis.

Reply:
Created a new issue for re-thinking / refactoring the factory pattern used across the codebase, see: #182 #182

Reply:
Created a new issue for re-thinking / refactoring the factory pattern used across the codebase, see: #182